agent/agentbootstrap/bootstrap.go

```diff
@@ -391,6 +411,45 @@ func initBootstrapMachine(
 	return m, nil
 }
 
+// initControllerCloudService creates cloud service for controller service.
+func initControllerCloudService(
+	isCAAS bool, st *state.State, args InitializeStateParams,
+) error {
+	if !isCAAS {
+		// IAAS does have controller cloud service for now.
+		logger.Debugf("no cloud service needs to be created for IAAS model")
+		return nil
+	}
+
+	env, err := getEnviron(isCAAS, args.ControllerModelConfig, args)
+	if err != nil {
+		return errors.Annotate(err, "getting environ")
+	}
+
+	broker, ok := env.(caas.ServiceGetterSetter)
+	if !ok {
+		// this should never happen.
+		return errors.New("CAAS environ is not a broker")
+	}
+	svc, err := broker.GetService(k8sprovider.JujuControllerStackName)
```

Comment:
How does this affect things like CDK and microk8s where you may have a DNS name for services, but that DNS only resolves *inside* the K8s (eg, your laptop probably doesn't point to the K8s' DNS)

That is one of the main reason that Juju has historically preferred IP addresses, because DNS is only reliable if it is globally resolvable DNS (and things like MAAS DNS suffer from the same problem.

I know that we do have some support for controllers being advertised at DNS names (it is how we integrate with JAAS). And we do things like negotiate Let's Encrypt to sign our certificates rather than just self-signed.

Reply:
The idea is we send the public service IP address to JUju clients (which is what this code is getting).
Internally we want to use something else, and ostensibly there's an internal cluster DNS and a naming scheme https://kubernetes.io/docs/concepts/services-networking/dns-pod-service. But experiments have shown that not everything behaves as documented at the moment.